files/aws_lifecycle_hooks/tools.py

```diff
@@ -1,62 +1,104 @@
 #!/usr/bin/env python
-'''
-File managed by puppet in module aws_lifecycle_hooks
-'''
+"""File managed by puppet in module aws_lifecycle_hooks."""
 import json
 import functools
 import urllib.request
 import urllib.error
 import typing
 import yaml
 import attr
+import re
 
 from exceptions import ParsingError
 
 
 @attr.s
 class VolumeAttachment:
+    """Utility class."""
+
     volume_id = attr.ib()
     device_name = attr.ib()
 
 
+def metadata_version() -> str:
+    """AWS Metadata version string."""
+    return '2020-10-27'
+
+
+@functools.lru_cache(maxsize=1)
+def get_metadata_token() -> str:
+    """
+    Instance Metadata Service Version 2 (IMDSv2).
+
+    https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+    Token fetching does not seem to work with versioned requests. Gives HTTP 403
+    """
+    headers = {'X-aws-ec2-metadata-token-ttl-seconds': 60}
+    req = urllib.request.Request(
+        url='http://169.254.169.254/latest/api/token',
+        headers=headers,
+        method='PUT',
+    )
+    with urllib.request.urlopen(req) as resp:
+        response_data = resp.read().decode("utf-8")
+    return response_data
+
+
+@functools.lru_cache(maxsize=1)
+def get_metadata(key: str) -> str:
+    """Fetch meta-data with latest version by key."""
+    headers = {'X-aws-ec2-metadata-token': get_metadata_token()}
+    req = urllib.request.Request(
+        url="http://169.254.169.254/{version}/{key}".format(
+            version=metadata_version(),
+            key=key,
+        ),
+        headers=headers,
+    )
+    try:
+        with urllib.request.urlopen(req) as resp:
+            response_data = resp.read().decode("utf-8")
+        return response_data
+    except urllib.error.HTTPError as e:
+        if e.status == 404:
+            print("Key '{key}' not found in meta-data service.".format(key=key))
```

Comment:
in this branch, you return an implicit None. Is that the intended behaviour?
If so, you should update the type hint to return `typing.Optional[str]`, or equivalently `typing.Union[None,str]`